### PR TITLE
[API] Add methods to add & remove smelting and anvil recipes during runtime

### PR DESCRIPTION
**Pull request by Mohron**

### CODE REVIEW

**Comment by Mohron** [comment]
> mezz> Mohron, how are you going about removing existing smelting and anvil recipes that are added by JEI internally?

I actually don't know. Will that even be possible?

**Comment by mezz** [comment]
You can get the recipe wrapper by creating a focus and looking it up with `IRecipeRegistry`. Is that good enough? I'm not sure there's a better way to do it.

File: src/main/java/mezz/jei/recipes/RecipeRegistry.java [hunk 1]
```diff
@@ -385,6 +388,28 @@ public void addSmeltingRecipe(List<ItemStack> inputs, ItemStack output) {
 	}
 
 	@Override
+	public IRecipeWrapper createSmeltingRecipe(List<ItemStack> inputs, ItemStack output) {
+		ErrorUtil.checkNotEmpty(inputs, "inputs");
+		Preconditions.checkNotNull(output, "null output");
```
Comment (mezz) [inline src/main/java/mezz/jei/recipes/RecipeRegistry.java]: There is `ErrorUtil.checkNotEmpty` for `ItemStack`, please use it here.

File: src/main/java/mezz/jei/recipes/RecipeRegistry.java [hunk 2]
```diff
@@ -385,6 +388,28 @@ public void addSmeltingRecipe(List<ItemStack> inputs, ItemStack output) {
...
+
+		SmeltingRecipe smeltingRecipe = new SmeltingRecipe(inputs, output);
+		addRecipe(smeltingRecipe);
```
Comment (mezz) [inline src/main/java/mezz/jei/recipes/RecipeRegistry.java]: This method should not add recipe, it should be a separate step.

File: src/main/java/mezz/jei/recipes/RecipeRegistry.java [hunk 3]
```diff
@@ -385,6 +388,28 @@ public void addSmeltingRecipe(List<ItemStack> inputs, ItemStack output) {
...
+
+	@Override
+	public IRecipeWrapper createAnvilRecipe(ItemStack leftInput, List<ItemStack> rightInputs, List<ItemStack> outputs) {
+		ErrorUtil.checkNotEmpty(leftInput);
+		ErrorUtil.checkNotEmpty(rightInputs, "rightInputs");
+		ErrorUtil.checkNotEmpty(outputs, "outputs");
+		Preconditions.checkArgument(rightInputs.size() == outputs.size(), "Input and output sizes must match.");
+
+		AnvilRecipeWrapper anvilRecipeWrapper = new AnvilRecipeWrapper(leftInput, rightInputs, outputs);
+		addRecipe(Collections.singletonList(anvilRecipeWrapper), VanillaRecipeCategoryUid.ANVIL);
```
Comment (mezz) [inline src/main/java/mezz/jei/recipes/RecipeRegistry.java]: This method should not add the recipe, it should be a separate step.

File: src/api/java/mezz/jei/api/IRecipeRegistry.java [hunk 4]
```diff
@@ -124,10 +124,34 @@
...
+	 * By default, all smelting recipes from {@link FurnaceRecipes#smeltingList} are already added by JEI.
+	 *
+	 * @param inputs the list of possible inputs to rotate through
+	 * @param output the output
+	 * @since JEI 4.3.3
+	 */
+	IRecipeWrapper createSmeltingRecipe(List<ItemStack> inputs, ItemStack output);
+
+	/**
+	 * Add a new anvil recipe while the game is running.
```
Comment (mezz) [inline src/api/java/mezz/jei/api/IRecipeRegistry.java]: Should be "Create a new anvil recipe."

File: src/api/java/mezz/jei/api/IRecipeRegistry.java [hunk 5]
```diff
@@ -124,10 +124,34 @@
 	 * By default, all smelting recipes from {@link FurnaceRecipes#smeltingList} are already added by JEI.
 	 *
 	 * @since JEI 4.2.7
+	 * @deprecated since JEI 4.3.3. Use {@link #createSmeltingRecipe(List, ItemStack)}
 	 */
+	@Deprecated
 	void addSmeltingRecipe(List<ItemStack> inputs, ItemStack output);
 
 	/**
+	 * Add a new smelting recipe while the game is running.
```
Comment (mezz) [inline src/api/java/mezz/jei/api/IRecipeRegistry.java]: Should be "Create a new smelting recipe."

**Review by mezz — CHANGES_REQUESTED** [review]

File: src/api/java/mezz/jei/api/IRecipeRegistry.java [hunk 6]
```diff
@@ -124,10 +124,34 @@
...
+	 * By default, all smelting recipes from {@link FurnaceRecipes#smeltingList} are already added by JEI.
```
Comment (Mohron) [inline src/api/java/mezz/jei/api/IRecipeRegistry.java]: @mezz do you want this line to stay?

Comment (mezz) [reply src/api/java/mezz/jei/api/IRecipeRegistry.java]: Yep

File: src/api/java/mezz/jei/api/IRecipeRegistry.java (hunk 5 above)
Comment (Mohron) [reply src/api/java/mezz/jei/api/IRecipeRegistry.java]: Should I add:
`Use {@link #addRecipe(IRecipeWrapper, String)} to add the recipe while the game is running.`
To each method's Doc tags as well?

Comment (mezz) [reply src/api/java/mezz/jei/api/IRecipeRegistry.java]: That's good, yeah.

File: src/main/java/mezz/jei/recipes/RecipeRegistry.java [hunk 7]
```diff
@@ -385,6 +388,24 @@ public void addSmeltingRecipe(List<ItemStack> inputs, ItemStack output) {
 	}
 
 	@Override
+	public IRecipeWrapper createSmeltingRecipe(List<ItemStack> inputs, ItemStack output) {
+		ErrorUtil.checkNotEmpty(inputs, "inputs");
+		Preconditions.checkNotNull(output);
```
Comment (mezz) [inline src/main/java/mezz/jei/recipes/RecipeRegistry.java]: There is `ErrorUtil.checkNotEmpty` for `ItemStack`, please use it here.

**Review by mezz — CHANGES_REQUESTED** [review]

**Comment by mezz** [comment]
Looks good, there is still one thing left to fix.

**Comment by Mohron** [comment]
Ya I missed the part where i was using something completely different (from the old API method) so I just removed the string param.

**Review by mezz — APPROVED** [review]

**Pull request merged**